sampler.go

```diff
@@ -199,45 +200,74 @@ func (s *rateLimitingSampler) Equal(other Sampler) bool {
 // The probabilisticSampler is given higher priority when tags are emitted, ie. if IsSampled() for both
 // samplers return true, the tags for probabilisticSampler will be used.
 type GuaranteedThroughputProbabilisticSampler struct {
-	probabilisticSampler Sampler
-	lowerBoundSampler    Sampler
-	tags                 []Tag
-	samplingRate         float64
-	lowerBound           float64
+	probabilisticSampler  Sampler
+	lowerBoundSampler     Sampler
+	upperBoundRateLimiter utils.RateLimiter
+	tags                  []Tag
+	samplingRate          float64
+	lowerBound            float64
+	upperBound            float64
 }
 
 // NewGuaranteedThroughputProbabilisticSampler returns a delegating sampler that applies both
 // probabilisticSampler and rateLimitingSampler.
 func NewGuaranteedThroughputProbabilisticSampler(
-	lowerBound, samplingRate float64,
+	lowerBound, samplingRate, upperBound float64,
 ) (*GuaranteedThroughputProbabilisticSampler, error) {
-	tags := []Tag{
-		{key: SamplerTypeTagKey, value: SamplerTypeLowerBound},
-		{key: SamplerParamTagKey, value: samplingRate},
-	}
 	probabilisticSampler, err := NewProbabilisticSampler(samplingRate)
 	if err != nil {
 		return nil, err
 	}
-	return &GuaranteedThroughputProbabilisticSampler{
-		probabilisticSampler: probabilisticSampler,
-		lowerBoundSampler:    NewRateLimitingSampler(lowerBound),
-		tags:                 tags,
-		samplingRate:         samplingRate,
-		lowerBound:           lowerBound,
-	}, nil
+	s := &GuaranteedThroughputProbabilisticSampler{
+		probabilisticSampler:  probabilisticSampler,
+		lowerBoundSampler:     NewRateLimitingSampler(lowerBound),
+		upperBoundRateLimiter: utils.NewRateLimiter(upperBound, upperBound),
+		samplingRate:          samplingRate,
+		lowerBound:            lowerBound,
+		upperBound:            upperBound,
+	}
+	s.setTags()
```

Comment:
I would rather have another helper function:
```
func (s *GuaranteedThroughputProbabilisticSampler) setSamplingRate(double) {
    if s.samplingRate != samplingRate {
        s.samplingRate = samplingRate
        s.probabilisticSampler = NewProbabilisticSampler(samplingRate)
        s.setTags() // may not need to be a function anymore if this is the only place
    }
}
```

we're doing this in several places, always together.